### PR TITLE
Add pl-sketch element

### PR DESCRIPTION
**Pull request by firasm**
A continuation of #2686. All credit to [kevin88tu](https://github.com/kevin88tu) for the initial code, and [nicknytko](https://github.com/nicknytko) for the reviews.

I'd love to see some rudimentary sketching capabilities added as an element, so upon advice of Matt and Nathan during PL office hours, I decided to extract out the element functionality.

I still think (as mentioned in that PR) using something more feature packed and more recently supported like [excalidraw](https://github.com/excalidraw/excalidraw) would be better, but this is a great starting point

### CODE REVIEW

**Comment by firasm** [comment]
Huzzah - looks like I was correctly able to extract all required components of the element.

Screenshot 2023-08-25 at 9 47 14 AM

File: apps/prairielearn/elements/pl-sketch/pl-sketch.mustache [hunk 1]
```diff
@@ -0,0 +1,50 @@
+<div class="pl-sketch" id="pl-sketch-{{uuid}}">
+    <div class="pl-sketch-controls" style="width: {{width}}px">
+        <div class="pl-sketch-control-group">
+            <div class="pl-sketch-control" id="undo"><div class="fa fa-undo fa-2x"></div></div>
+            <div class="pl-sketch-control" id="redo"><div class="fa fa-redo fa-2x"></div></div>
+            <div class="pl-sketch-control" id="pen"><div class="fa fa-pen fa-2x"></div></div>
+            <div class="pl-sketch-control" id="eraser"><div class="fa fa-eraser fa-2x"></div></div>
+        </div>
```
Comment (jonatanschroeder) [inline apps/prairielearn/elements/pl-sketch/pl-sketch.mustache:8]: The use of `id` here is probably a bad idea. Any other element that uses the same id (including more than one sketch in the same page) would cause an issue here.

File: apps/prairielearn/elements/pl-sketch/pl-sketch.mustache [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        let strokes = {};
+        let sketches_string = `{{{sketchpad_json}}}`;
+        if (sketches_string.length != 0) {
+            sketches_string = sketches_string.replace(/'/g, '!@#$%');
+            sketches_string = sketches_string.replace(/"/g, '\'');
+            sketches_string = sketches_string.replace(/!@#\$%/g, '"');
+            sketches = JSON.parse(sketches_string);
+            strokes = sketches['sketches'];
+            strokes = strokes.replace(/'/g, '"');
+            strokes = JSON.parse(strokes);
```
Comment (jonatanschroeder) [inline apps/prairielearn/elements/pl-sketch/pl-sketch.mustache:43]: This looks sketchy (pun intended)... I think it makes more sense to save the data as the value of the hidden input (with double braces mustache, so it's escaped), then use the input's value as the source to parse the JSON content. Also, you probably want to move this to the JS file to reduce the amount of JS in the HTML file.

File: apps/prairielearn/elements/pl-sketch/pl-sketch.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    if data["panel"] == "question":
+        if len(data["raw_submitted_answers"]) == 0:
+            skp_json = ""
+        else:
+            skp_json = data["raw_submitted_answers"]
+
+        html_params = {
+            "uuid": pl.get_uuid(),
+            "sketchpad_json": skp_json,
+            "test_elem": "hello world",
```
Comment (jonatanschroeder) [inline apps/prairielearn/elements/pl-sketch/pl-sketch.py:23]: leftover from testing?

File: apps/prairielearn/elements/pl-sketch/pl-sketch.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+        }
+
+        with open("pl-sketch.mustache", "r") as f:
+            return chevron.render(f, html_params).strip()
+    else:
+        return ""
+
+
+def grade(element_html, data):
+    data["partial_scores"]["pl-sketch-unique-name"] = {"score": 1, "weight": 1}
```
Comment (jonatanschroeder) [inline apps/prairielearn/elements/pl-sketch/pl-sketch.py:35]: This won't work if you have multiple sketch elements. Also, if the element doesn't provide any grading, is there a reason to compute a grade?

File: package.json [hunk 5]
```diff
@@ -42,5 +42,8 @@
   "engines": {
     "node": ">=16.0.0"
   },
-  "packageManager": "yarn@3.5.0"
+  "packageManager": "yarn@3.5.0",
+  "dependencies": {
+    "sketchpad": "^0.1.0"
```
Comment (jonatanschroeder) [inline package.json:47]: This dependency should be added to apps/prairielearn, not the root.

**Review by jonatanschroeder — COMMENTED** [review]
This is just some initial thoughts from a first skim through the content.

File: apps/prairielearn/elements/pl-sketch/pl-sketch.mustache (hunk 1 above)
Comment (nwalters512) [reply apps/prairielearn/elements/pl-sketch/pl-sketch.mustache:8]: As for what to do instead: either use a distinct class on each of these and then select based on that, or include the `uuid` in the ID to ensure they're unique on the page.

File: apps/prairielearn/elements/pl-sketch/pl-sketch.py
Comment (eliotwrobson) [inline apps/prairielearn/elements/pl-sketch/pl-sketch.py:1]: This file is pretty old, needs to have type annotations added and `pl-sketch` added to the `pyproject.toml` file.

File: apps/prairielearn/elements/pl-sketch/pl-sketch.mustache [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+                    <div class="pl-sketch-circle" style="background-color: #795548"></div>
+                    <div class="pl-sketch-circle" style="background-color: #E0E0E0"></div>
+                    <div class="pl-sketch-circle" style="background-color: #616161"></div>
+                    <div class="pl-sketch-circle" style="background-color: #000000"></div>
+                </div>
+            </div>
+        </div>
+    </div>
+    <canvas class="pl-sketch-canvas" id="pl-sketch-canvas-{{uuid}}" data-width="80%" data-height="80%"></canvas>
+    <input type="hidden" id="pl-sketch-input-{{uuid}}" name="sketches" value="" />
```
Comment (nwalters512) [inline apps/prairielearn/elements/pl-sketch/pl-sketch.mustache:30]: Related to Jonatan's comment in the Python, using a static `name` here means that having multiple sketch elements in one question will be impossible. This element should accept an `answers-name` attribute, and that attribute should be used for the name here.

For instance, see `pl-string-input`:

https://github.com/PrairieLearn/PrairieLearn/blob/dbffba773e4f598496a7e410dfe51aa71d8f1533/apps/prairielearn/elements/pl-string-input/pl-string-input.py#L60

https://github.com/PrairieLearn/PrairieLearn/blob/dbffba773e4f598496a7e410dfe51aa71d8f1533/apps/prairielearn/elements/pl-string-input/pl-string-input.mustache#L19

File: apps/prairielearn/elements/pl-sketch/pl-sketch.py [hunk 7]
```diff
@@ -0,0 +1,36 @@
+import chevron
+import lxml.html
+import prairielearn as pl
+
+
+def prepare(element_html, data):
+    return data
+
+
```
Comment (nwalters512) [inline apps/prairielearn/elements/pl-sketch/pl-sketch.py:9]: This is a no-op and should be removed.

File: apps/prairielearn/elements/pl-sketch/pl-sketch.py [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+def render(element_html, data):
+    element = lxml.html.fragment_fromstring(element_html)
+    width = pl.get_integer_attrib(element, "width", 600)
+    height = pl.get_integer_attrib(element, "height", 700)
+    if data["panel"] == "question":
```
Comment (eliotwrobson) [inline apps/prairielearn/elements/pl-sketch/pl-sketch.py:14]: You can unindent this by returning immediately if the panel not `"question"`.

File: apps/prairielearn/elements/pl-sketch/pl-sketch.py (hunk 4 above)
Comment (eliotwrobson) [reply apps/prairielearn/elements/pl-sketch/pl-sketch.py:35]: Agree, you can just delete the grade function completely.

File: docs/elements.md [hunk 9]
```diff
@@ -55,6 +55,7 @@ images, files, and code display. The following **decorative** elements are avail
 - [`pl-dataframe`](#pl-dataframe-element): Display DataFrames with various options.
 - [`pl-drawing`](#pl-drawing-element): Creates an image from pre-defined
   collection of graphic objects
+- [`pl-sketch`](#pl-sketch-element): Freely draw anything using a sketchpad.
```
Comment (eliotwrobson) [inline docs/elements.md:58]: There was a PR a while ago that alphabetized the elements. Should move the sketch docs to maintain that ordering.

File: docs/elements/pl-sketch.png
Comment (eliotwrobson) [inline docs/elements/pl-sketch.png:1]: I think we stopped using images as part of the docs a while ago? I think it actually might make sense to keep in this case however.

File: exampleCourse/questions/element/sketch/question.html [hunk 10]
```diff
@@ -0,0 +1,5 @@
+<pl-question-panel>
+  <p>Example use of a pl-sketch.</p>
+</pl-question-panel>
+
+<pl-sketch width=500 height=600></pl-sketch>
```
Comment (eliotwrobson) [inline exampleCourse/questions/element/sketch/question.html:5]: Maybe make this wider? At least center it.

File: apps/prairielearn/elements/pl-sketch/pl-sketch.mustache (hunk 1 above)
Comment (eliotwrobson) [reply apps/prairielearn/elements/pl-sketch/pl-sketch.mustache:8]: Like with the `id` above, you need to add the `uuid` to the end.

File: apps/prairielearn/elements/pl-sketch/pl-sketch.py [hunk 11]
```diff
@@ -0,0 +1,36 @@
+import chevron
+import lxml.html
+import prairielearn as pl
+
+
+def prepare(element_html, data):
```
Comment (eliotwrobson) [inline apps/prairielearn/elements/pl-sketch/pl-sketch.py:6]: I think this function can just be deleted. Also no need to return data (is modified in-place).

**Review by eliotwrobson — COMMENTED** [review]
Some early comments. Looking decent so far.

File: apps/prairielearn/elements/pl-sketch/pl-sketch.mustache [hunk 12]
```diff
@@ -0,0 +1,50 @@
+<div class="pl-sketch" id="pl-sketch-{{uuid}}">
+    <div class="pl-sketch-controls" style="width: {{width}}px">
+        <div class="pl-sketch-control-group">
+            <div class="pl-sketch-control" id="undo"><div class="fa fa-undo fa-2x"></div></div>
+            <div class="pl-sketch-control" id="redo"><div class="fa fa-redo fa-2x"></div></div>
+            <div class="pl-sketch-control" id="pen"><div class="fa fa-pen fa-2x"></div></div>
+            <div class="pl-sketch-control" id="eraser"><div class="fa fa-eraser fa-2x"></div></div>
```
Comment (nwalters512) [inline apps/prairielearn/elements/pl-sketch/pl-sketch.mustache:7]: Although I acknowledge that this element isn't going to be particularly accessible to keyboard users since there's no straightforward way for them to draw, we still should not be using a `<div>` as a clickable thing. A `<button>` is absolutely the way to go.

File: apps/prairielearn/elements/pl-sketch/pl-sketch.py [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+            "sketchpad_json": skp_json,
+            "test_elem": "hello world",
+            "width": width,
+            "height": height,
+        }
+
+        with open("pl-sketch.mustache", "r") as f:
+            return chevron.render(f, html_params).strip()
+    else:
+        return ""
```
Comment (nwalters512) [inline apps/prairielearn/elements/pl-sketch/pl-sketch.py:31]: Why doesn't this element render itself in the submission panel?

File: docs/elements.md (hunk 9 above)
Comment (nwalters512) [inline docs/elements.md:58]: Let's maintain the alphabetical ordering of this list.

File: docs/elements/pl-sketch.png
Comment (nwalters512) [inline docs/elements/pl-sketch.png:1]: It might be nice to show this element with some input? Maybe something like the molecular diagram you put in the PR description?

File: exampleCourse/questions/element/sketch/question.html (hunk 10 above)
Comment (nwalters512) [inline exampleCourse/questions/element/sketch/question.html:5]:
```suggestion
<pl-sketch width="500" height="600"></pl-sketch>
```

File: apps/prairielearn/elements/pl-sketch/info.json [hunk 14]
```diff
@@ -0,0 +1,8 @@
+{
+  "controller": "pl-sketch.py",
+  "dependencies": {
+    "nodeModulesScripts": ["sketchpad/scripts/sketchpad.js"],
```
Comment (nwalters512) [inline apps/prairielearn/elements/pl-sketch/info.json:4]: This element appears to be almost entirely unmaintained; it hasn't seen a release in 7 years and appears to be somewhat limited in what it can do. Did you look into any other options?

That said, the source code isn't too long (~300 lines) and is MIT licensed, so we should be covered if we decide we need to fork it to support more use cases.

File: apps/prairielearn/elements/pl-sketch/pl-sketch.css [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+}
+
+.pl-sketch-controls {
+  height: 40px;
+  display: flex;
+  align-items: center;
+  justify-content: center;
+  background: #fafafa;
+  box-shadow: 0px 0px 2px #757575;
+  z-index: 2;
```
Comment (nwalters512) [inline apps/prairielearn/elements/pl-sketch/pl-sketch.css:15]: Where possible, we prefer using Bootstrap classes to style elements. That means `d-flex`, `align-items-center`, and so on.

File: apps/prairielearn/elements/pl-sketch/pl-sketch.mustache [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+                    <div class="pl-sketch-circle" style="background-color: #4CAF50"></div>
+                    <div class="pl-sketch-circle" style="background-color: #795548"></div>
+                    <div class="pl-sketch-circle" style="background-color: #E0E0E0"></div>
+                    <div class="pl-sketch-circle" style="background-color: #616161"></div>
+                    <div class="pl-sketch-circle" style="background-color: #000000"></div>
+                </div>
+            </div>
+        </div>
+    </div>
+    <canvas class="pl-sketch-canvas" id="pl-sketch-canvas-{{uuid}}" data-width="80%" data-height="80%"></canvas>
```
Comment (nwalters512) [inline apps/prairielearn/elements/pl-sketch/pl-sketch.mustache:29]: What do `data-width` and `data-height` do?

File: apps/prairielearn/elements/pl-sketch/pl-sketch.py (hunk 11 above)
Comment (eliotwrobson) [reply apps/prairielearn/elements/pl-sketch/pl-sketch.py:6]: Wait, actually you should check for having the desired attributes on this element in prepare.

File: apps/prairielearn/elements/pl-sketch/pl-sketch.py (hunk 7 above)
Comment (nwalters512) [reply apps/prairielearn/elements/pl-sketch/pl-sketch.py:9]: As Eliot pointed out in his review, you should indeed be validating attributes here.

**Comment by jonatanschroeder** [comment]
> I still think (as mentioned in that PR) using something more feature packed and more recently supported like [excalidraw](https://github.com/excalidraw/excalidraw) would be better, but this is a great starting point

Excalidraw, from what I could check, depends on react, so it's probably not an option.

File: package.json (hunk 5 above)
Comment (firasm) [reply package.json:47]: agreed, removed and added there.

File: apps/prairielearn/elements/pl-sketch/pl-sketch.py (hunk 4 above)
Comment (firasm) [reply apps/prairielearn/elements/pl-sketch/pl-sketch.py:35]: Agreed, I thought it was weird that it would always give a score of 1 - was on my list of things to remove.

File: apps/prairielearn/elements/pl-sketch/pl-sketch.py (hunk 3 above)
Comment (firasm) [reply apps/prairielearn/elements/pl-sketch/pl-sketch.py:23]: thanks

File: apps/prairielearn/elements/pl-sketch/pl-sketch.py (hunk 8 above)
Comment (firasm) [reply apps/prairielearn/elements/pl-sketch/pl-sketch.py:14]: good suggestion, thanks!

File: docs/elements.md (hunk 9 above)
Comment (firasm) [reply docs/elements.md:58]: got it, will do.

File: exampleCourse/questions/element/sketch/question.html (hunk 10 above)
Comment (firasm) [reply exampleCourse/questions/element/sketch/question.html:5]: For now, I made it wider... I wish there was a reasonable way to span the entire question panel without causing downstream issues (students on different device resolutions submitting things that look entirely different.)

File: apps/prairielearn/elements/pl-sketch/pl-sketch.py
Comment (firasm) [reply apps/prairielearn/elements/pl-sketch/pl-sketch.py:1]: I added pl-sketch to pyproject.toml in a previous commit. 

Still need to do the type annotations...

File: apps/prairielearn/elements/pl-sketch/info.json (hunk 14 above)
Comment (firasm) [reply apps/prairielearn/elements/pl-sketch/info.json:4]: I'm not married to this package at all and have not done any research into other options... will try and look around but I don't have much capacity to make comprehensive changes at the moment, but on the other hand I'd also love additional functionality (annotating on pictures, button to wipe the canvas, download the sketch as a png, instructor annotations during manual grading, etc... the list can be long!).

**Comment by firasm** [comment]
Thanks for the reviews everyone! I've addressed and resolved a few of the comments, the remaining ones will take a bit more work than I have time for right this second, but I do agree they need to be resolved before proceeding. If someone has extra cycles, please feel free to continue... hopefully I come back and don't forget about this!

File: apps/prairielearn/elements/pl-sketch/pl-sketch.mustache (hunk 1 above)
Comment (elisacarrillo) [reply apps/prairielearn/elements/pl-sketch/pl-sketch.mustache:8]: Update:
New lines are as follows - 
```
<div class="pl-sketch-control" id="undo-{{uuid}}"><div class="fa fa-undo fa-2x"></div></div>
<div class="pl-sketch-control" id="redo-{{uuid}}"><div class="fa fa-redo fa-2x"></div></div>
<div class="pl-sketch-control" id="pen-{{uuid}}"><div class="fa fa-pen fa-2x"></div></div>
<div class="pl-sketch-control" id="eraser-{{uuid}}"><div class="fa fa-eraser fa-2x"></div></div>
```

And can be referenced in pl-sketch.js as '#pen-' + uuid, '#undo-' + uuid,...

File: apps/prairielearn/elements/pl-sketch/pl-sketch.js [hunk 17]
```diff
@@ -0,0 +1,93 @@
...
+    var sketchpadObject = {};
+    sketchpadObject['width'] = width;
+    sketchpadObject['height'] = height;
+    sketchpadObject['element'] = element;
+    let sketches_string = $('#pl-sketch-input-' + uuid).val();
+    console.log(sketches_string.length);
+    if (sketches_string.length != 0) {
+        sketches_string = sketches_string.replace(/'/g, '!@#$%');
+        sketches_string = sketches_string.replace(/"/g, '\'');
+        sketches_string = sketches_string.replace(/!@#\$%/g, '"');
```
Comment (jonatanschroeder) [inline apps/prairielearn/elements/pl-sketch/pl-sketch.js:51]: Why is this needed? My understanding is that this string is always saved with `JSON.stringify`, so why do you need to replace quotes?

File: apps/prairielearn/elements/pl-sketch/pl-sketch.js [hunk 18]
```diff
@@ -0,0 +1,93 @@
...
+    }
+  },
+
+  setupSketchpad: function (width, height, uuid) {
+    console.log('#undo-'+uuid);
+    let element = '#pl-sketch-canvas-' + uuid;
+    var sketchpadObject = {};
+    sketchpadObject['width'] = width;
+    sketchpadObject['height'] = height;
+    sketchpadObject['element'] = element;
```
Comment (jonatanschroeder) [inline apps/prairielearn/elements/pl-sketch/pl-sketch.js:45]:
```suggestion
    const sketchpadObject = { width, height, element };
```
There are other simplifications that can be done in other places in the code, such as using `const` instead of `let`, but this is one example.

File: apps/prairielearn/elements/pl-sketch/pl-sketch.js [hunk 19]
```diff
@@ -0,0 +1,93 @@
...
+    if (toolName === this.tool) {
+      return;
+    }
+    console.log(toolName);
+    
+    if (toolName === 'pen') {
+      size = 5;
+      this.tool = 'pen';
+      $('#pen').css('background', '#C0C0C0');
+      $('#eraser').css('background', '#FAFAFA');
```
Comment (jonatanschroeder) [inline apps/prairielearn/elements/pl-sketch/pl-sketch.js:18]: This still relies on the unique ID, which is no longer used.

File: apps/prairielearn/elements/pl-sketch/pl-sketch.js [hunk 20]
```diff
@@ -0,0 +1,93 @@
...
+        sketches_string = sketches_string.replace(/!@#\$%/g, '"');
+        let sketches = JSON.parse(sketches_string);
+        let strokes = sketches['sketches'];
+        strokes = strokes.replace(/'/g, '"');
+        strokes = JSON.parse(strokes);
+
+        sketchpadObject['strokes'] = strokes;
+        sketchpadObject['undoHistory'] = sketches['undoHistory'];
+    }
+    this.sketchpad = new Sketchpad(sketchpadObject);
```
Comment (jonatanschroeder) [inline apps/prairielearn/elements/pl-sketch/pl-sketch.js:60]: `this` in this case is `window.PLSketchAPI`, which is unique within the page. This will give you again issues with multiple sketches in the same question. I would encourage you take a look at the strategy used in other elements like [`pl-file-editor`](https://github.com/PrairieLearn/PrairieLearn/blob/master/apps/prairielearn/elements/pl-file-editor/pl-file-editor.js) or [`pl-order-blocks`](https://github.com/PrairieLearn/PrairieLearn/blob/master/apps/prairielearn/elements/pl-order-blocks/pl-order-blocks.js) for ways to address this kind of process.

File: apps/prairielearn/elements/pl-sketch/pl-sketch.mustache [hunk 21]
```diff
@@ -0,0 +1,43 @@
+<div class="pl-sketch" id="pl-sketch-{{uuid}}">
+    <div class="pl-sketch-controls" style="width: {{width}}px">
+        <div class="pl-sketch-control-group">
+            <div class="pl-sketch-control" id="undo-{{uuid}}"><div class="fa fa-undo fa-2x"></div></div>
+            <div class="pl-sketch-control" id="redo-{{uuid}}"><div class="fa fa-redo fa-2x"></div></div>
+            <div class="pl-sketch-control" id="pen-{{uuid}}"><div class="fa fa-pen fa-2x"></div></div>
+            <div class="pl-sketch-control" id="eraser-{{uuid}}"><div class="fa fa-eraser fa-2x"></div></div>
```
Comment (jonatanschroeder) [inline apps/prairielearn/elements/pl-sketch/pl-sketch.mustache:7]: Personally, I'm not a huge fan of having a ton of elements with uuid-based IDs. One option you can use here is to have classes, with something like:
```suggestion
            <div class="pl-sketch-control pl-sketch-undo"><div class="fa fa-undo fa-2x"></div></div>
            <div class="pl-sketch-control pl-sketch-redo"><div class="fa fa-redo fa-2x"></div></div>
            <div class="pl-sketch-control pl-sketch-pen"><div class="fa fa-pen fa-2x"></div></div>
            <div class="pl-sketch-control pl-sketch-eraser"><div class="fa fa-eraser fa-2x"></div></div>
```
This should allow you to use `.pl-sketch-pen`, for example, to refer to any pen element, or `#pl-sketch-{{uuid}} .pl-sketch-pen` to refer to a specific pen control.

File: apps/prairielearn/elements/pl-sketch/pl-sketch.mustache [hunk 22]
```diff
@@ -0,0 +1,43 @@
...
+                    <div class="pl-sketch-circle" style="background-color: #795548"></div>
+                    <div class="pl-sketch-circle" style="background-color: #E0E0E0"></div>
+                    <div class="pl-sketch-circle" style="background-color: #616161"></div>
+                    <div class="pl-sketch-circle" style="background-color: #000000"></div>
+                </div>
+            </div>
+        </div>
+    </div>
+    <canvas class="pl-sketch-canvas" id="pl-sketch-canvas-{{uuid}}" data-width="80%" data-height="80%"></canvas>
+    <input type="hidden" id="pl-sketch-input-{{uuid}}" name="sketches-{{uuid}}" value="{{{sketchpad_json}}}" />
```
Comment (jonatanschroeder) [inline apps/prairielearn/elements/pl-sketch/pl-sketch.mustache:30]: You probably shouldn't use the UUID for a name here. Since UUID is created on render, this means every time you submit a new UUID is used, which means the name will be different in every call. You could change the Python code to generate UUID in `prepare` and save it in `params`, but I believe it makes more sense for you to have an `answers-name` parameter, for consistency with other elements that use the same model.

File: apps/prairielearn/elements/pl-sketch/pl-sketch.py [hunk 23]
```diff
@@ -0,0 +1,31 @@
...
+    width = pl.get_integer_attrib(element, "width", 600)
+    height = pl.get_integer_attrib(element, "height", 700)
+
+    if data["panel"] != "question":
+        return ""
+
+    if len(data["raw_submitted_answers"]) == 0:
+        skp_json = ""
+    else:
+        skp_json = data["raw_submitted_answers"]
```
Comment (jonatanschroeder) [inline apps/prairielearn/elements/pl-sketch/pl-sketch.py:21]: The content of the sketchpad will not be directly in `data["raw_submitted_answers"]`, but in `data["raw_submitted_answers"][answers_name]` (see other comment).

File: docker-compose.yml [hunk 24]
```diff
@@ -1,6 +1,7 @@
 version: '3.8'
 services:
   pl:
+    platform: linux/amd64
```
Comment (jonatanschroeder) [inline docker-compose.yml:4]: This is unrelated. Was it added by mistake?

File: docs/elements.md [hunk 25]
```diff
@@ -1442,6 +1443,38 @@ The HTML inside the inner `pl-variable` tag is used as the description. If the t
...
+
+| Attribute | Type    | Default | Description                  |
+| --------- | ------- | ------- | ---------------------------- |
+| `width`   | integer | 500     | Width of sketchpad element.  |
+| `height`  | integer | 600     | Height of sketchpad element. |
+
+#### Details
+
+Any submission will be marked correct, as there is no way to automatically grade a sketch.
+
```
Comment (jonatanschroeder) [inline docs/elements.md:1471]: Do you actually want to grade it? I don't think it should be (see elements like `pl-rich-text-editor` which are also used for input but don't directly affect grading). Even if you do, note that you should consider questions that contain other elements as well as a sketch, and in that case the score of other elements should also be considered. In any case, there is no grade function in your element.

File: docker-compose.yml (hunk 24 above)
Comment (elisacarrillo) [reply docker-compose.yml:4]: Hey yes sorry I added this when I was trying to fix my docker issues, just wanted to comment so I know this issue has been resolved :) Thanks for these comments!

**Comment by jonatanschroeder** [comment]
I was looking a bit around, and I believe that [js-draw](https://www.npmjs.com/package/js-draw) may be a better option for this kind of element. It is actively maintained, and has way better features. If you're still interested in a drawing/sketching-like object, I suggest to close this PR and open a new one based on js-draw.

**Comment by firasm** [comment]
I talked to Matt about this the other day in PL office hours and he said Excalidraw would also be totally fine to use (despite React). 

I'm hopeful I'll be hiring a student to explore a few options. 

Thanks for the js-draw reference too!

**Comment by nwalters512** [comment]
@firasm given that it sounds like you've found a student to work on a `pl-excalidraw` element, do you still wish to keep this PR open, or can we close it?

**Comment by firasm** [comment]
Yep We can close this.
